writer.go

```diff
@@ -19,28 +17,45 @@ type Writer struct {
 	framer    Framer
 	formatter Formatter
 
+	mu   sync.RWMutex // guards conn
 	conn serverConn
 }
 
+// getConn provides access to the internal conn, protected by a mutex. The
+// conn is threadsafe, so it can be used while unlocked, but we want to avoid
+// race conditions on grabbing a reference to it.
+func (w *Writer) getConn() serverConn {
+	w.mu.RLock()
+	defer w.mu.RUnlock()
+	return w.conn
+}
+
+// setConn updates the internal conn, protected by a mutex.
+func (w *Writer) setConn(c serverConn) {
+	w.mu.Lock()
+	w.conn = c
+	w.mu.Unlock()
+}
+
 // connect makes a connection to the syslog server.
-// It must be called with w.mu held.
-func (w *Writer) connect() (err error) {
-	if w.conn != nil {
+func (w *Writer) connect() (serverConn, error) {
+	conn := w.getConn()
+	if conn != nil {
 		// ignore err from close, it makes sense to continue anyway
-		w.conn.close()
-		w.conn = nil
+		conn.close()
+		w.setConn(nil)
 	}
 
-	var conn serverConn
 	var hostname string
+	var err error
 	dialer := w.getDialer()
 	conn, hostname, err = dialer.Call()
 	if err == nil {
-		w.conn = conn
+		w.setConn(conn)
 		w.hostname = hostname
 	}
 
-	return
+	return conn, err
```

Comment:
It isn't explicitly specified in the comments but is rather "expected to be implied" that if `err != nil` then the `conn` (or any other return values) are invalid and shouldn't be used.

It would perhaps be more accurate to explicitly return `nil` in that case, but frankly the only thing that would do is cause a nil pointer dereference if somewhere else in the code was incorrectly not checking `err` before using the `conn`. I'm not seeing a _huge_ benefit to that, but I'm happy to change it if we want to insist.